src/wp-includes/html-api/class-wp-xml-decoder.php

```diff
@@ -0,0 +1,263 @@
+<?php
+
+/**
+ * XML API: WP_XML_Decoder class
+ *
+ * Decodes spans of raw text found inside XML content.
+ *
+ * @package WordPress
+ * @subpackage HTML-API
+ * @since WP_VERSION
+ */
+class WP_XML_Decoder {
+
+	/**
+	 * Decodes a span of XML text.
+	 *
+	 * Example:
+	 *
+	 *     '©' = WP_XML_Decoder::decode( 'data', '&copy;' );
+	 *
+	 * @since WP_VERSION
+	 *
+	 * @access private
+	 *
+	 * @param string $text    Text document containing span of text to decode.
+	 * @return string Decoded UTF-8 string.
+	 */
+	public static function decode( $text ) {
+		$decoded = '';
+		$end     = strlen( $text );
+		$at      = 0;
+		$was_at  = 0;
+
+		while ( $at < $end ) {
+			$next_character_reference_at = strpos( $text, '&', $at );
+			if ( false === $next_character_reference_at || $next_character_reference_at >= $end ) {
+				break;
+			}
+			// $next_character_reference_at += 1;
+
+			/*
+			 * Capture all bytes that could form a character reference.
+			 *
+			 * This only supports:
+			 *
+			 * * The five mandated character references, that is &amp; &lt; &gt; &quot; &apos;
+			 * * Numeric character references, e.g. &#123; or &#x1A;
+			 *
+			 * XML grammar rule for parsing numeric references is:
+			 *
+			 *     [66] CharRef   ::= '&#' [0-9]+ ';' | '&#x' [0-9a-fA-F]+ ';' [WFC: Legal Character]
+			 *
+			 * See https://www.w3.org/TR/xml/#NT-CharRef
+			 */
+			$token_length = strspn(
+				$text,
+				'ampltgquos#xX0123456789bcdefABCDEF;',
+				$next_character_reference_at + 1,
+				/*
+				 * Limit the length of the token to avoid scanning the entire document in case
+				 * a semicolon is missing.
+				 *
+				 * The maximum supported code point is 10FFFF, which is 9 characters long when
+				 * represented as either decimal or hexadecimal numeric character reference entity.
+				 * Technically, you can also add zeros to the front of the entity, which makes the
+				 * string longer, for example &#00000010FFFF;
+				 *
+				 * We limit this scan to 30 characters, which allows twenty zeros at the front.
+				 */
+				30
+			);
```

Comment:
in this case I don't think we want a limit on these. for one, XML requires the trailing semicolon, and for two, we don't want to add rules to the spec and accidentally let something slip in, for example, by prepending a long reference in the front.

I can clean this up here, but I think it's important we not mush together all of the characters. something like this…

```php
if ( '#' === $text[ $next_character_reference_at + 1 ] ) {
	$is_hex = 'x' === $text[ $next_character_reference_at + 2 ] || 'X' === $text[ … ];
	$zeros_start_at  = $next_character_reference_at + 3 + ( $is_hex ? 1 : 0 );
	$zeros_length    = strspn( $text, '0', $zeros_start_at );
	$digits_start_at = $zeros_start_at + $zeros_length;
	$digit_chars     = $is_hex ? '0123456789abcdefABCDEF' : '0123456789';
	$digits_length   = strspn( $text, $digit_chars, $digits_start_at );
	$semicolon_at    = $digits_start_at + $digits_length;

	// Must be followed by a semicolon.
	if ( ';' !== $text[ $semicolon_at ] ) {
		return false;
	}

	// Null bytes cannot be encoded in XML.
	if ( 0 === $digits_length ) {
		return false;
	}

	/*
	 * Must encode a valid Unicode code point.
	 * (Avoid parsing more than is necessary).
	 */
	$max_digits = $is_hex ? 6 : 7;
	if ( $digits_length > $max_digits ) {
		return false;
	}

	$base       = $is_hex ? 16 : 10;
	$code_point = intval( substr( $text, $digits_start_at, $digits_length ), $base );
	if ( if_allowable_code_point( $code_point ) ) {
		$decoded .= WP_HTML_Decoder::code_point_to_utf8_bytes( $code_point );
		continue;
	}

	return false;
}

// Must be a named character reference.
$name_starts_at = $next_character_reference_at + 1;

$standard_entities = array(
	'amp;'  => '&',
	'apos;' => "'",
	'gt;'   => '>',
	'lt;'   => '<',
	'quot;' => '"',
);

foreach ( $standard_entities as $name => $replacement ) {
	if ( substr_compare( $text, $name, $name_starts_at, strlen( $name ) ) ) {
		$decoded .= $replacement;
		continue;
	}
}
```

granted, we want to perform all length checks to avoid crashing, but I think we _have_ to scan the entire thing to avoid mis-parsing and security issues

Reply:
Oh, this won't just continue after seeing something like `&#x000000000000000000000000000000000000000000000000000000000000000000000123`. If we don't get a valid reference within those 30 characters that's a parse failure.